### PR TITLE
Mark/1488 customize histogram calculations

### PR DESCRIPTION
**Pull request by markccchiang**
**Description**

This customized the histogram calculations as described in #1488. The companion [backend PR](https://github.com/CARTAvis/carta-backend/pull/1254) and [protobuf PR](https://github.com/CARTAvis/carta-protobuf/pull/83) are shown on the hyperlinks.

**Checklist**

For linked issues (if there are):
- [x] assignee and label added
- [x] ZenHub issue connection, board status, and estimate updated

For the pull request:
- [x] reviewers and assignee added
- [x] ZenHub estimate, milestone, and release (if needed) added
- [x] e2e test passing / ~corresponding fix added~
- [x] changelog updated ~/ no changelog update needed~
- [x] protobuf updated to the latest dev commit ~/ no protobuf update needed~
- [x] `BackendService` unchanged ~/ `BackendService` changed and corresponding ICD test fix added~

### CODE REVIEW

**Comment by kswang1029** [comment]
this looks good. please proceed code review.

**Review by kswang1029 — APPROVED** [review]
this looks fine. no regression is seen from e2e tests.

**Comment by acdo2002** [comment]
@markccchiang 
Because the `src/services/BackendService` is the same to the current dev one, I think the final checking item can change to "BackendService unchanged", then tick the checkbox. 

The setHistogramRequirements request in the backendService is direct adopt the protobuf defined format (i.e. CARTA.ISetHistogramRequirements), so it does not require additional modification in the BackendService.ts in the frontend code.

**Comment by markccchiang** [comment]
I am fixing a bug found by @YuHsuan-Hwang. And also changing the codes according to suggestions. Will push new commits when they are done.

File: src/components/Histogram/HistogramComponent.tsx [hunk 1]
```diff
@@ -37,6 +37,7 @@ export class HistogramComponent extends React.Component<WidgetProps> {
     }
 
     private cachedFrame: FrameStore;
+    private curLinePlotProps: LinePlotComponentProps;
```
Comment (YuHsuan-Hwang) [inline src/components/Histogram/HistogramComponent.tsx]: I prefer we rename it to `currentLinePlotProps` for clarity.

File: src/components/Histogram/HistogramComponent.tsx [hunk 2]
```diff
@@ -53,31 +54,25 @@ export class HistogramComponent extends React.Component<WidgetProps> {
...
-                return null;
-            }
-            const stokesIndex = this.widgetStore.effectiveFrame.polarizationInfo.findIndex(polarization => polarization.replace("Stokes ", "") === coordinate.slice(0, coordinate.length - 1));
-            const stokes = stokesIndex >= this.widgetStore.effectiveFrame.frameInfo.fileInfoExtended.stokes ? this.widgetStore.effectiveFrame.polarizations[stokesIndex] : stokesIndex;
-            const regionHistogramData = regionMap.get(stokes === -1 ? this.widgetStore.effectiveFrame.requiredStokes : stokes);
-            if (!regionHistogramData) {
-                return null;
-            }
-            return regionHistogramData.histograms;
+        // Check whether the hostogram data matchs the wieget's configuration
```
Comment (YuHsuan-Hwang) [inline src/components/Histogram/HistogramComponent.tsx]: typo: widget

File: src/components/Histogram/HistogramSettingsPanelComponent/HistogramConfigPanelComponent.tsx [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+import {observer} from "mobx-react";
+
+import {SafeNumericInput} from "components/Shared";
+import {HistogramWidgetStore} from "stores/Widgets";
+
+@observer
+export class HistogramConfigPanelComponent extends React.Component<{widgetStore: HistogramWidgetStore}> {
+    private static readonly BINS_LOWER_BOUND = 2;
+
+    private currentMaxNumBins: number;
```
Comment (YuHsuan-Hwang) [inline src/components/Histogram/HistogramSettingsPanelComponent/HistogramConfigPanelComponent.tsx]: I suggest we remove `currentMaxNumBins` and `maxNumBins` in the component and just use the variable `widgetStore.maxNumBins` from the store for the max bin num input. `this.currentMaxNumBins > HistogramConfigPanelComponent.BINS_LOWER_BOUND` check can be moved to `onMaxNumBinsChanged`.

File: src/components/Histogram/HistogramSettingsPanelComponent/HistogramConfigPanelComponent.tsx [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+    private changeNumBinsHandler = (numBins: number) => {
+        this.widgetStore.setNumBins(numBins);
+    };
+
+    private onResetConfig = () => {
+        this.widgetStore.onResetConfig();
+
+        // Reset the maximum number of bins for the bins slider and its filler
+        const newMaxNumBins = this.widgetStore.currentNumBins * 2;
```
Comment (YuHsuan-Hwang) [inline src/components/Histogram/HistogramSettingsPanelComponent/HistogramConfigPanelComponent.tsx]: Shall we reset this to the default bin number * 2? If `currentNumBins` was changed to a small value, we might reset it to a value smaller than the default bin number.

**Review by YuHsuan-Hwang — CHANGES_REQUESTED** [review]
The changes are looking good! I have some feedback for the details:

File: src/components/Histogram/HistogramComponent.tsx (hunk 2 above)
Comment (YuHsuan-Hwang) [reply src/components/Histogram/HistogramComponent.tsx]: @markccchiang wideget -> widget
The rest of changes all look good to me 👍

Comment (markccchiang) [reply src/components/Histogram/HistogramComponent.tsx]: Oops, sorry for missing it!

**Review by YuHsuan-Hwang — APPROVED** [review]

**Review by kswang1029 — APPROVED** [review]
looking good and no regression from e2e tests. 👍

**Comment by YuHsuan-Hwang** [comment]
Added fix for #2150.

**Pull request merged**